templates/commands/render.go

```diff
@@ -355,6 +351,89 @@ func (c *RenderCommand) checkUnknownInputs(spec *model.Spec) []string {
 	return unknownInputs
 }
 
+// resolveInputs combines flags, user prompts, and defaults to get the full set
+// of template inputs.
+func (c *RenderCommand) resolveInputs(ctx context.Context, spec *model.Spec) error {
+	if unknownInputs := c.checkUnknownInputs(spec); len(unknownInputs) > 0 {
+		return fmt.Errorf("unknown input(s): %s", strings.Join(unknownInputs, ", "))
+	}
+
+	if c.flags.Prompt {
+		isATTY := (c.Stdin() == os.Stdin && isatty.IsTerminal(os.Stdin.Fd()))
+		if !isATTY {
+			return fmt.Errorf("the flag --prompt was provided, but standard input is not a terminal")
+		}
+
+		if err := c.promptForInputs(ctx, spec); err != nil {
+			return err
+		}
+	} else {
+		c.collapseDefaultInputs(spec)
+		if missing := c.checkInputsMissing(spec); len(missing) > 0 {
+			return fmt.Errorf("missing input(s): %s", strings.Join(missing, ", "))
+		}
+	}
+
+	return nil
+}
+
+// promptForInputs looks for template inputs that were not provided on the
+// command line and prompts the user for them. This mutates c.flags.Inputs.
+//
+// This must only be called when the user specified --prompt and the input is a
+// terminal (or in a test).
+func (c *RenderCommand) promptForInputs(ctx context.Context, spec *model.Spec) error {
+	for _, i := range spec.Inputs {
+		if _, ok := c.flags.Inputs[i.Name.Val]; ok {
+			// Don't prompt if the cmdline had an --input for this key.
+			continue
+		}
+		sb := &strings.Builder{}
+		tw := tabwriter.NewWriter(sb, 8, 0, 2, ' ', 0)
+		fmt.Fprintf(tw, "\nInput name:\t%s", i.Name.Val)
+		fmt.Fprintf(tw, "\nDescription:\t%s", i.Desc.Val)
+
+		if i.Default != nil {
+			defaultStr := i.Default.Val
+			if defaultStr == "" {
+				// When empty string is the default, print it differently so
+				// the user can actually see what's happening.
+				defaultStr = `""`
+			}
+			fmt.Fprintf(tw, "\nDefault:\t%s", defaultStr)
+			tw.Flush()
+
+			// Design note: we can't treat "empty string" as meaning "use
+			// the default value" because empty string is a legitimate input
+			// value.
+
+			fmt.Fprintf(sb, "\n\nAccept the (d)efault or (p)rovide your own value? (d/p): ")
```

Comment:
I'm not so sure about this, it doesn't seem very clean. But I can't think of a better way. LMK.

Reply:
Can you provide sample output of a couple prompts in the description?

Reply:
Done, the part I'm concerned about is the `Accept the (d)efault or (p)rovide your own value? (d/p)`.

Reply:
I wonder if we should just use defaults if they are available.

Reply:
What are the situations where the empty string is a valid value **and** the default value is not the empty string?

Reply:
> I wonder if we should just use defaults if they are available.

That feels risky to me. The user might be surprised that they silently got a value that doesn't work for them.

Reply:
> What are the situations where the empty string is a valid value **and** the default value is not the empty string?

I think that's what I needed to hear. In the cases where empty string is a valid input, it can also be the default. I'll update the code.

Reply:
Updated, LMK what you think. The example prompts in the PR description have also been updated.